source/auth_signer.c

```diff
@@ -0,0 +1,193 @@
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "auth.h"
+
+#include "http.h"
+
+#include <aws/auth/signable.h>
+#include <aws/auth/signer.h>
+
+static const char *s_capsule_name_signer = "aws_signer";
+
+/* Signer capsule contains raw aws_signer struct. There is no intermediate binding struct. */
+
+/* Runs when GC destroys the capsule containing the binding */
+static void s_signer_capsule_destructor(PyObject *py_capsule) {
+    struct aws_signer *signer = PyCapsule_GetPointer(py_capsule, s_capsule_name_signer);
+    aws_signer_destroy(signer);
+}
+
+struct aws_signer *aws_py_get_signer(PyObject *py_signer) {
+    struct aws_signer *native = NULL;
+
+    PyObject *py_capsule = PyObject_GetAttrString(py_signer, "_binding");
+    if (py_capsule) {
+        native = PyCapsule_GetPointer(py_capsule, s_capsule_name_signer);
+        Py_DECREF(py_capsule);
+    }
+
+    return native;
+}
+
+PyObject *aws_py_signer_new_aws(PyObject *self, PyObject *args) {
+    (void)self;
+
+    if (!PyArg_ParseTuple(args, "")) {
+        return NULL;
+    }
+
+    struct aws_signer *signer = aws_signer_new_aws(aws_py_get_allocator());
+    if (!signer) {
+        return PyErr_AwsLastError();
+    }
+
+    /* From hereon, we need to clean up if errors occur. */
+
+    PyObject *py_capsule = PyCapsule_New(signer, s_capsule_name_signer, s_signer_capsule_destructor);
+    if (!py_capsule) {
+        aws_signer_destroy(signer);
+        return NULL;
+    }
+
+    return py_capsule;
+}
+
+/* Object that stays alive for duration async signing operation */
+struct async_signing_data {
+    PyObject *py_signer;
+    PyObject *py_http_request;
+    PyObject *py_signing_config;
+    PyObject *py_on_complete;
+    struct aws_signable *signable;
+};
+
+static void s_async_signing_data_destroy(struct async_signing_data *async_data) {
+    if (async_data) {
+        Py_XDECREF(async_data->py_signer);
+        Py_XDECREF(async_data->py_http_request);
+        Py_XDECREF(async_data->py_signing_config);
+        Py_XDECREF(async_data->py_on_complete);
+        aws_signable_destroy(async_data->signable);
+    }
+}
+
+static void s_signing_complete(struct aws_signing_result *signing_result, int error_code, void *userdata) {
+    struct async_signing_data *async_data = userdata;
+
+    /*************** GIL ACQUIRE ***************/
+    PyGILState_STATE state = PyGILState_Ensure();
+
+    if (error_code) {
+        goto done;
+    }
+
+    struct aws_http_message *http_request = aws_py_get_http_message(async_data->py_http_request);
+    if (!http_request) {
+        error_code = AWS_ERROR_UNKNOWN;
+        goto done;
+    }
+
+    if (aws_apply_signing_result_to_http_request(http_request, aws_py_get_allocator(), signing_result)) {
```

Comment:
should we do this stuff prior to acquiring the GIL?

Reply:
Yeah, we could do that